# Init

In [1]:
import pandas as pd
import numpy as np

seed = 18022004
np.random.seed(seed)

In [2]:
original_df = pd.read_parquet("official_original.parquet")
original_df

id          fromLib                             toLib  \
0          0          ant:ant                org.apache.ant:ant   
1          1          ant:ant                org.apache.ant:ant   
2          2          ant:ant                org.apache.ant:ant   
7          7          ant:ant                org.apache.ant:ant   
8          8          ant:ant                org.apache.ant:ant   
...      ...              ...                               ...   
14329  14329  xstream:xstream  com.thoughtworks.xstream:xstream   
14330  14330  xstream:xstream  com.thoughtworks.xstream:xstream   
14331  14331  xstream:xstream  com.thoughtworks.xstream:xstream   
14332  14332  xstream:xstream  com.thoughtworks.xstream:xstream   
14333  14333  xstream:xstream  com.thoughtworks.xstream:xstream   

                       repoName                                  fileName  \
0           bobmcwhirter_drools                                   pom.xml   
1            wocommunity_wolips               woproject-ant-tasks/pom.xml   
2                   teiid_teiid                                   pom.xml   
7                 mulesoft_mule                        standalone/pom.xml   
8                 mulesoft_mule          distributions/standalone/pom.xml   
...                         ...                                       ...   
14329  kiegroup_kogito-runtimes                       drools-core/pom.xml   
14330        sakaiproject_sakai            gradebook/service/impl/pom.xml   
14331        sakaiproject_sakai      gradebook/service/sakai-pack/pom.xml   
14332            apache_shindig  1.0.x-incubating/java/social-api/pom.xml   
14333            apache_shindig                  1.0.x-incubating/pom.xml   

                                    startCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      endCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      startCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.apache.ant:ant\n-ant:ant   
...                                                  ...   
14329  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14330  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14331  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14332  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14333  +com.thoughtworks.xstream:xstream\n-xstream:xs...   

                                        endCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.a

In [3]:
original_df["repoName"].nunique()

1480

## Insight about log test library

In [4]:
tolib_test_df = original_df[original_df["toLib"].str.contains("unit|test")]

tolib_test_df["repoName"].nunique()

231

In [5]:
fromlib_test_df = original_df[original_df["fromLib"].str.contains("unit|test")]
fromlib_test_df["repoName"].nunique()

260

In [6]:
tolib_test_df

id                              fromLib  \
45        45                              ant:ant   
731      731      com.google.android:android-test   
1125    1125  com.jayway.restassured:rest-assured   
1358    1358         com.sun.jersey:jersey-bundle   
3084    3084                    httpunit:httpunit   
...      ...                                  ...   
14308  14308                      xmlunit:xmlunit   
14309  14309                      xmlunit:xmlunit   
14310  14310                      xmlunit:xmlunit   
14311  14311                      xmlunit:xmlunit   
14312  14312                      xmlunit:xmlunit   

                                                   toLib  \
45                              org.apache.ant:ant-junit   
731           com.google.testing.compile:compile-testing   
1125   org.glassfish.jersey.test-framework.providers:...   
1358   org.glassfish.jersey.test-framework.providers:...   
3084                   net.sourceforge.htmlunit:htmlunit   
...                                                  ...   
14308                         org.xmlunit:xmlunit-legacy   
14309                         org.xmlunit:xmlunit-legacy   
14310                         org.xmlunit:xmlunit-legacy   
14311                         org.xmlunit:xmlunit-legacy   
14312                         org.xmlunit:xmlunit-legacy   

                              repoName  \
45                  wocommunity_wolips   
731                  hoko_hoko-android   
1125       meridor_perspective-backend   
1358   opensecuritycontroller_osc-core   
3084                    apache_archiva   
...                                ...   
14308             opendaylight_netconf   
14309             opendaylight_netconf   
14310           opendaylight_yangtools   
14311           opendaylight_yangtools   
14312          opendaylight_controller   

                                                fileName  \
45                           woproject-ant-tasks/pom.xml   
731                                              pom.xml   
1125                                             pom.xml   
1358                                  osc-common/pom.xml   
3084   archiva-modules/archiva-web/archiva-webdav/pom...   
...                                                  ...   
14308         netconf/netconf-notifications-impl/pom.xml   
14309              netconf/sal-netconf-connector/pom.xml   
14310                     yang/yang-model-export/pom.xml   
14311                        yang/yang-data-impl/pom.xml   
14312  opendaylight/md-sal/sal-clustering-commons/pom...   

                                    startCommit  \
45     1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
731    c59afec9c1d13c21c118373a5e972805a2db5649   
1125   6595e1a76e4179f0cd732d65819c7902c4ac3c60   
1358   1de6cb2c58551212d572f33f2cad36dc533c92b7   
3084   4825f8eb88ef35ca265903515ae715014bb4d597   
...                                         ...   
14308  652dada8b53d21c3eefb409b18365bbc29560afe   
14309  652dada8b53d21c3eefb409b18365bbc29560afe   
14310  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14311  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14312  e8c4d500b42c1b9b5d49b2f72a0c1cd9bcc42261   

                                      endCommit  \
45     1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
731    c59afec9c1d13c21c118373a5e972805a2db5649   
1125   6595e1a76e4179f0cd732d65819c7902c4ac3c60   
1358   1de6cb2c58551212d572f33f2cad36dc533c92b7   
3084   4825f8eb88ef35ca265903515ae715014bb4d597   
...                                         ...   
14308  652dada8b53d21c3eefb409b18365bbc29560afe   
14309  652dada8b53d21c3eefb409b18365bbc29560afe   
14310  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14311  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14312  e8c4d500b42c1b9b5d49b2f72a0c1cd9bcc42261   

                                      startCommitChanges  \
45     +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
731    +org.easytesting:fest-assert-core\n+com.square...   
1125   +org.glassfis

In [7]:
fromlib_test_df

id                          fromLib  \
731      731  com.google.android:android-test   
732      732  com.google.android:android-test   
733      733  com.google.android:android-test   
3084    3084                httpunit:httpunit   
3085    3085                httpunit:httpunit   
...      ...                              ...   
14308  14308                  xmlunit:xmlunit   
14309  14309                  xmlunit:xmlunit   
14310  14310                  xmlunit:xmlunit   
14311  14311                  xmlunit:xmlunit   
14312  14312                  xmlunit:xmlunit   

                                            toLib                 repoName  \
731    com.google.testing.compile:compile-testing        hoko_hoko-android   
732                   org.robolectric:robolectric        hoko_hoko-android   
733                   org.robolectric:robolectric             BUS-Team_bus   
3084            net.sourceforge.htmlunit:htmlunit           apache_archiva   
3085    org.apache.tomcat.embed:tomcat-embed-core           apache_archiva   
...                                           ...                      ...   
14308                  org.xmlunit:xmlunit-legacy     opendaylight_netconf   
14309                  org.xmlunit:xmlunit-legacy     opendaylight_netconf   
14310                  org.xmlunit:xmlunit-legacy   opendaylight_yangtools   
14311                  org.xmlunit:xmlunit-legacy   opendaylight_yangtools   
14312                  org.xmlunit:xmlunit-legacy  opendaylight_controller   

                                                fileName  \
731                                              pom.xml   
732                                              pom.xml   
733                                      android/pom.xml   
3084   archiva-modules/archiva-web/archiva-webdav/pom...   
3085   archiva-modules/archiva-web/archiva-webdav/pom...   
...                                                  ...   
14308         netconf/netconf-notifications-impl/pom.xml   
14309              netconf/sal-netconf-connector/pom.xml   
14310                     yang/yang-model-export/pom.xml   
14311                        yang/yang-data-impl/pom.xml   
14312  opendaylight/md-sal/sal-clustering-commons/pom...   

                                    startCommit  \
731    c59afec9c1d13c21c118373a5e972805a2db5649   
732    c59afec9c1d13c21c118373a5e972805a2db5649   
733    cb50f8e47cbbf550e2421c9be86f14a96acfea2b   
3084   4825f8eb88ef35ca265903515ae715014bb4d597   
3085   4825f8eb88ef35ca265903515ae715014bb4d597   
...                                         ...   
14308  652dada8b53d21c3eefb409b18365bbc29560afe   
14309  652dada8b53d21c3eefb409b18365bbc29560afe   
14310  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14311  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14312  e8c4d500b42c1b9b5d49b2f72a0c1cd9bcc42261   

                                      endCommit  \
731    c59afec9c1d13c21c118373a5e972805a2db5649   
732    c59afec9c1d13c21c118373a5e972805a2db5649   
733    cb50f8e47cbbf550e2421c9be86f14a96acfea2b   
3084   4825f8eb88ef35ca265903515ae715014bb4d597   
3085   4825f8eb88ef35ca265903515ae715014bb4d597   
...                                         ...   
14308  652dada8b53d21c3eefb409b18365bbc29560afe   
14309  652dada8b53d21c3eefb409b18365bbc29560afe   
14310  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14311  e1057eac909eb2b1968557a2d0ad4b8d631e1682   
14312  e8c4d500b42c1b9b5d49b2f72a0c1cd9bcc42261   

                                      startCommitChanges  \
731    +org.easytesting:fest-assert-core\n+com.square...   
732    +org.easytesting:fest-assert-core\n+com.square...   
733    +org.roboguice:roboguice\n+org.robolectric:rob...   
3084   +org.apache.tomcat:tomcat-juli\n+xerces:xerces...   
3085   +org.apache.tomcat:tomcat-juli\n+xerces:xerces...   
...                                                  ...   
14308      +org.xmlunit:xmlunit-legacy\n-xmlunit:xmlunit   
14309      +org.xmlunit:xmlunit-legacy\n-xmlunit:xmlunit 

In [8]:
print(tolib_test_df.index)

Index([   45,   731,  1125,  1358,  3084,  3088,  3089,  3090,  3091,  3092,
       ...
       14303, 14304, 14305, 14306, 14307, 14308, 14309, 14310, 14311, 14312],
      dtype='int64', length=1383)


In [9]:
print(fromlib_test_df.index)

Index([  731,   732,   733,  3084,  3085,  3086,  3087,  3088,  3089,  3090,
       ...
       14303, 14304, 14305, 14306, 14307, 14308, 14309, 14310, 14311, 14312],
      dtype='int64', length=1453)


In [10]:
print(len(set(tolib_test_df.index).intersection(set(fromlib_test_df.index))))
print(len(set(tolib_test_df.index).union(set(fromlib_test_df.index))))

1297
1539


In [11]:
1297 / len(tolib_test_df)

0.9378163412870572

In [12]:
original_df[original_df["id"] == 3084].iloc[0]

id                                                                 3084
fromLib                                               httpunit:httpunit
toLib                                 net.sourceforge.htmlunit:htmlunit
repoName                                                 apache_archiva
fileName              archiva-modules/archiva-web/archiva-webdav/pom...
startCommit                    4825f8eb88ef35ca265903515ae715014bb4d597
endCommit                      4825f8eb88ef35ca265903515ae715014bb4d597
startCommitChanges    +org.apache.tomcat:tomcat-juli\n+xerces:xerces...
endCommitChanges      +org.apache.tomcat:tomcat-juli\n+xerces:xerces...
startCommitMessage    remove use of old httpunit library\n\ngit-svn-...
endCommitMessage      remove use of old httpunit library\n\ngit-svn-...
startCommitTime                                              2014-01-06
endCommitTime                                                2014-01-06
Category                                                        

In [13]:
i = 3084

print(original_df.loc[i, "fromLib"])
print(original_df.loc[i, "toLib"])
print(original_df.loc[i, "repoName"])
print(original_df.loc[i, "fileName"])

httpunit:httpunit
net.sourceforge.htmlunit:htmlunit
apache_archiva
archiva-modules/archiva-web/archiva-webdav/pom.xml


In [14]:
i = 3085

print(original_df.loc[i, "fromLib"])
print(original_df.loc[i, "toLib"])
print(original_df.loc[i, "repoName"])
print(original_df.loc[i, "fileName"])

httpunit:httpunit
org.apache.tomcat.embed:tomcat-embed-core
apache_archiva
archiva-modules/archiva-web/archiva-webdav/pom.xml


In [15]:
i = 3086

print(original_df.loc[i, "fromLib"])
print(original_df.loc[i, "toLib"])
print(original_df.loc[i, "repoName"])
print(original_df.loc[i, "fileName"])

httpunit:httpunit
org.apache.tomcat.embed:tomcat-embed-logging-juli
apache_archiva
archiva-modules/archiva-web/archiva-webdav/pom.xml


In [16]:
i = 3087

print(original_df.loc[i, "fromLib"])
print(original_df.loc[i, "toLib"])
print(original_df.loc[i, "repoName"])
print(original_df.loc[i, "fileName"])

httpunit:httpunit
org.apache.tomcat:tomcat-juli
apache_archiva
archiva-modules/archiva-web/archiva-webdav/pom.xml


In [17]:
i = 3088

print(original_df.loc[i, "fromLib"])
print(original_df.loc[i, "toLib"])
print(original_df.loc[i, "repoName"])
print(original_df.loc[i, "fileName"])

httpunit:httpunit
org.httpunit:httpunit
webx_citrus
test/util/pom.xml


In [18]:
set(tolib_test_df.index) - set(fromlib_test_df.index)

{45,
 1125,
 1358,
 3223,
 3224,
 3225,
 4255,
 4256,
 4257,
 4258,
 4259,
 4260,
 4261,
 4262,
 4263,
 4264,
 4265,
 4266,
 4267,
 4268,
 4269,
 4270,
 4271,
 4272,
 4273,
 4274,
 4275,
 4276,
 4277,
 4278,
 4279,
 4280,
 4281,
 6432,
 9157,
 9158,
 9485,
 9577,
 9588,
 9657,
 9658,
 9659,
 9660,
 9661,
 9662,
 9663,
 9664,
 9673,
 9674,
 9675,
 9676,
 9677,
 9678,
 9679,
 9680,
 9681,
 9682,
 9683,
 9684,
 9685,
 9700,
 9701,
 9702,
 9703,
 9707,
 10216,
 10471,
 10506,
 10507,
 10509,
 10510,
 10512,
 10513,
 10581,
 10582,
 10593,
 10594,
 10595,
 11170,
 11173,
 11293,
 13785,
 13786,
 13787,
 13788,
 13789}

In [19]:
set(fromlib_test_df.index) - set(tolib_test_df.index)

{732,
 733,
 3085,
 3086,
 3087,
 3093,
 5250,
 5251,
 5252,
 5286,
 9211,
 9212,
 9213,
 9214,
 9215,
 9216,
 9217,
 9218,
 9219,
 9220,
 9221,
 9222,
 9223,
 9224,
 9225,
 9226,
 9227,
 9228,
 9229,
 9230,
 9231,
 9232,
 9233,
 9234,
 9235,
 9236,
 9237,
 9238,
 9239,
 9240,
 9241,
 9242,
 9243,
 9244,
 9245,
 9246,
 9247,
 9248,
 9249,
 9250,
 9251,
 9252,
 9253,
 9254,
 9255,
 9256,
 9257,
 9258,
 9259,
 9260,
 9261,
 9262,
 9263,
 9264,
 9265,
 9266,
 9267,
 9268,
 9269,
 9270,
 9271,
 9272,
 9273,
 9274,
 9275,
 9276,
 9277,
 9278,
 9279,
 9280,
 9281,
 9282,
 9283,
 9284,
 9285,
 9286,
 9287,
 9288,
 9289,
 9290,
 9291,
 9292,
 9293,
 9294,
 9295,
 9296,
 9297,
 9306,
 9307,
 9308,
 9309,
 9310,
 9311,
 9312,
 9313,
 9314,
 9315,
 9316,
 9317,
 9318,
 9319,
 9320,
 9321,
 9322,
 9323,
 9324,
 9326,
 9327,
 9328,
 9329,
 9920,
 10217,
 10218,
 10219,
 10220,
 10221,
 10222,
 10223,
 10359,
 10360,
 10377,
 10378,
 10379,
 10380,
 10381,
 10382,
 11267,
 11268,
 11269,
 11270,
 11

In [20]:
i = 9158
print(original_df.loc[i, "fromLib"])
print(original_df.loc[i, "toLib"])
print(original_df.loc[i, "repoName"])
print(original_df.loc[i, "fileName"])

org.dom4j:dom4j
org.xmlunit:xmlunit-matchers
apache_maven-archetype
archetype-common/pom.xml


***Cases that fromLib has different function than toLib***

- 1358

com.sun.jersey:jersey-bundle

org.glassfish.jersey.test-framework.providers:jersey-test-framework-provider-grizzly2

opensecuritycontroller_osc-core

osc-common/pom.xml

- 45
ant:ant

org.apache.ant:ant-junit

wocommunity_wolips

woproject-ant-tasks/pom.xml

- 3093

httpunit:httpunit

xerces:xercesImpl

apache_archiva

archiva-modules/archiva-web/archiva-webdav/pom.xml

- 5286

junit:junit

org.springframework.boot:spring-boot-starter-web

VIPJoey_doe

mmc-dubbo-provider/pom.xml

- 6432

net.htmlparser.jericho:jericho-html

net.sourceforge.htmlunit:htmlunit

apache_cxf-fediz

systests/tests/pom.xml

- 9158

org.dom4j:dom4j

org.xmlunit:xmlunit-matchers

apache_maven-archetype

archetype-common/pom.xml

***Testing framework that not has test|unit***
com.jayway.restassured:rest-assured

org.robolectric:robolectric

io.rest-assured:rest-assured

jmock:jmock

org.mockito:mockito-core

org.assertj:assertj-core

In [21]:
tolib_log_df = original_df[original_df["toLib"].str.contains("log|slf4j")]

tolib_log_df["repoName"].nunique()

426

In [22]:
fromlib_log_df = original_df[original_df["toLib"].str.contains("log|slf4j")]

fromlib_log_df["repoName"].nunique()

426

In [23]:
tolib_log_df

id                                            fromLib  \
180      180                      ch.qos.logback:logback-access   
181      181                      ch.qos.logback:logback-access   
182      182                      ch.qos.logback:logback-access   
183      183                      ch.qos.logback:logback-access   
184      184                      ch.qos.logback:logback-access   
...      ...                                                ...   
13789  13789                             org.slf4j:slf4j-simple   
13836  13836  org.springframework.boot:spring-boot-starter-l...   
13837  13837  org.springframework.boot:spring-boot-starter-l...   
13838  13838  org.springframework.boot:spring-boot-starter-l...   
13839  13839  org.springframework.boot:spring-boot-starter-l...   

                                     toLib  \
180     org.apache.logging.log4j:log4j-api   
181     org.apache.logging.log4j:log4j-api   
182    org.apache.logging.log4j:log4j-core   
183    org.apache.logging.log4j:log4j-core   
184     org.apache.logging.log4j:log4j-jcl   
...                                    ...   
13789           org.spf4j:spf4j-slf4j-test   
13836       ch.qos.logback:logback-classic   
13837       ch.qos.logback:logback-classic   
13838   org.apache.logging.log4j:log4j-api   
13839      commons-logging:commons-logging   

                                     repoName  \
180               flipkart-incubator_Poseidon   
181               flipkart-incubator_Poseidon   
182               flipkart-incubator_Poseidon   
183               flipkart-incubator_Poseidon   
184               flipkart-incubator_Poseidon   
...                                       ...   
13789                        zolyfarkas_spf4j   
13836               HotelsDotCom_circus-train   
13837               HotelsDotCom_circus-train   
13838                         eclipse_hawkbit   
13839  spring-cloud_spring-cloud-cloudfoundry   

                                          fileName  \
180                                        pom.xml   
181                              container/pom.xml   
182                                        pom.xml   
183                              container/pom.xml   
184                                        pom.xml   
...                                            ...   
13789                            spf4j-zel/pom.xml   
13836         circus-train-s3-mapreduce-cp/pom.xml   
13837                    circus-train-core/pom.xml   
13838         examples/hawkbit-example-app/pom.xml   
13839  spring-cloud-cloudfoundry-discovery/pom.xml   

                                    startCommit  \
180    2552502ec262f35c76d95856dbdec902884a9567   
181    2552502ec262f35c76d95856dbdec902884a9567   
182    2552502ec262f35c76d95856dbdec902884a9567   
183    2552502ec262f35c76d95856dbdec902884a9567   
184    2552502ec262f35c76d95856dbdec902884a9567   
...                                         ...   
13789  f5afbdf7dd26acf089e44f30c24e9122465f9d9b   
13836  407ec931245c21503c0f13216129b114e8b3bd35   
13837  407ec931245c21503c0f13216129b114e8b3bd35   
13838  9d6527a3e29567c0c0b55951121d1356a65d7a3b   
13839  1a55cc1c716d7fca26adeddea5a70575ad60f005   

                                      endCommit  \
180    2552502ec262f35c76d95856dbdec902884a9567   
181    2552502ec262f35c76d95856dbdec902884a9567   
182    2552502ec262f35c76d95856dbdec902884a9567   
183    2552502ec262f35c76d95856dbdec902884a9567   
184    2552502ec262f35c76d95856dbdec902884a9567   
...                                         ...   
13789  f5afbdf7dd26acf089e44f30c24e9122465f9d9b   
13836  407ec931245c21503c0f13216129b114e8b3bd35   
13837  407ec931245c21503c0f13216129b114e8b3bd35   
13838  9d6527a3e29567c0c0b55951121d1356a65d7a3b   
13839  1a55cc1c716d7fca26adeddea5a70575ad60f005   

                                      startCommitChanges  \
180    +org.apache.logging.log4j:log4j-jcl\n+org.apac...   
181    +org.apache.logging.log4j:log4j-core\n+org.apa...   
182    +

In [24]:
fromlib_log_df

id                                            fromLib  \
180      180                      ch.qos.logback:logback-access   
181      181                      ch.qos.logback:logback-access   
182      182                      ch.qos.logback:logback-access   
183      183                      ch.qos.logback:logback-access   
184      184                      ch.qos.logback:logback-access   
...      ...                                                ...   
13789  13789                             org.slf4j:slf4j-simple   
13836  13836  org.springframework.boot:spring-boot-starter-l...   
13837  13837  org.springframework.boot:spring-boot-starter-l...   
13838  13838  org.springframework.boot:spring-boot-starter-l...   
13839  13839  org.springframework.boot:spring-boot-starter-l...   

                                     toLib  \
180     org.apache.logging.log4j:log4j-api   
181     org.apache.logging.log4j:log4j-api   
182    org.apache.logging.log4j:log4j-core   
183    org.apache.logging.log4j:log4j-core   
184     org.apache.logging.log4j:log4j-jcl   
...                                    ...   
13789           org.spf4j:spf4j-slf4j-test   
13836       ch.qos.logback:logback-classic   
13837       ch.qos.logback:logback-classic   
13838   org.apache.logging.log4j:log4j-api   
13839      commons-logging:commons-logging   

                                     repoName  \
180               flipkart-incubator_Poseidon   
181               flipkart-incubator_Poseidon   
182               flipkart-incubator_Poseidon   
183               flipkart-incubator_Poseidon   
184               flipkart-incubator_Poseidon   
...                                       ...   
13789                        zolyfarkas_spf4j   
13836               HotelsDotCom_circus-train   
13837               HotelsDotCom_circus-train   
13838                         eclipse_hawkbit   
13839  spring-cloud_spring-cloud-cloudfoundry   

                                          fileName  \
180                                        pom.xml   
181                              container/pom.xml   
182                                        pom.xml   
183                              container/pom.xml   
184                                        pom.xml   
...                                            ...   
13789                            spf4j-zel/pom.xml   
13836         circus-train-s3-mapreduce-cp/pom.xml   
13837                    circus-train-core/pom.xml   
13838         examples/hawkbit-example-app/pom.xml   
13839  spring-cloud-cloudfoundry-discovery/pom.xml   

                                    startCommit  \
180    2552502ec262f35c76d95856dbdec902884a9567   
181    2552502ec262f35c76d95856dbdec902884a9567   
182    2552502ec262f35c76d95856dbdec902884a9567   
183    2552502ec262f35c76d95856dbdec902884a9567   
184    2552502ec262f35c76d95856dbdec902884a9567   
...                                         ...   
13789  f5afbdf7dd26acf089e44f30c24e9122465f9d9b   
13836  407ec931245c21503c0f13216129b114e8b3bd35   
13837  407ec931245c21503c0f13216129b114e8b3bd35   
13838  9d6527a3e29567c0c0b55951121d1356a65d7a3b   
13839  1a55cc1c716d7fca26adeddea5a70575ad60f005   

                                      endCommit  \
180    2552502ec262f35c76d95856dbdec902884a9567   
181    2552502ec262f35c76d95856dbdec902884a9567   
182    2552502ec262f35c76d95856dbdec902884a9567   
183    2552502ec262f35c76d95856dbdec902884a9567   
184    2552502ec262f35c76d95856dbdec902884a9567   
...                                         ...   
13789  f5afbdf7dd26acf089e44f30c24e9122465f9d9b   
13836  407ec931245c21503c0f13216129b114e8b3bd35   
13837  407ec931245c21503c0f13216129b114e8b3bd35   
13838  9d6527a3e29567c0c0b55951121d1356a65d7a3b   
13839  1a55cc1c716d7fca26adeddea5a70575ad60f005   

                                      startCommitChanges  \
180    +org.apache.logging.log4j:log4j-jcl\n+org.apac...   
181    +org.apache.logging.log4j:log4j-core\n+org.apa...   
182    +

In [25]:
print(tolib_log_df.index)
print(fromlib_log_df.index)

Index([  180,   181,   182,   183,   184,   185,   186,   187,   188,   189,
       ...
       13784, 13785, 13786, 13787, 13788, 13789, 13836, 13837, 13838, 13839],
      dtype='int64', length=4023)
Index([  180,   181,   182,   183,   184,   185,   186,   187,   188,   189,
       ...
       13784, 13785, 13786, 13787, 13788, 13789, 13836, 13837, 13838, 13839],
      dtype='int64', length=4023)


In [26]:
for lib in original_df["toLib"].sample(n=500):
    print(lib)

org.awaitility:awaitility
log4j:log4j
org.apache.logging.log4j:log4j-api
org.postgresql:postgresql
org.mockito:mockito-core
com.fasterxml.jackson.core:jackson-core
org.junit.platform:junit-platform-launcher
org.apache.commons:commons-lang3
com.fasterxml.jackson.core:jackson-annotations
org.kevoree:org.kevoree.log
org.easytesting:fest-assert
org.slf4j:slf4j-api
org.glassfish.jersey.core:jersey-server
org.apache.httpcomponents:httpclient
org.apache.commons:commons-configuration2
org.junit.jupiter:junit-jupiter-api
com.sun.jersey:jersey-core
org.eclipse.jetty:jetty-servlets
org.mybatis:mybatis
ch.qos.logback:logback-classic
org.apache.commons:commons-collections4
javax.ws.rs:javax.ws.rs-api
org.mortbay.jetty:jetty
org.eclipse.jetty:jetty-webapp
org.apache.commons:commons-io
log4j:log4j
com.carrotsearch:hppc
com.fasterxml.jackson.core:jackson-core
com.jayway.restassured:rest-assured
io.cucumber:cucumber-java
com.fasterxml.jackson.core:jackson-annotations
org.glassfish.jersey.core:jersey-cl

In [27]:
hard_remove_log_test_repos = (
    set(original_df["repoName"].unique())
    - set(fromlib_log_df["repoName"].unique())
    - set(tolib_log_df["repoName"].unique())
    - set(fromlib_test_df["repoName"].unique())
    - set(tolib_test_df["repoName"].unique())
)
len(hard_remove_log_test_repos)

844

In [28]:
# com.jayway.restassured:rest-assured
# org.robolectric:robolectric
# io.rest-assured:rest-assured
# jmock:jmock
# org.mockito:mockito-core
# org.assertj:assertj-core

In [28]:
fromlib_addition_test_df = original_df[
    original_df["fromLib"].str.contains("rest-assured|robolectric|mock|assert")
]
fromlib_addition_test_df

id                                fromLib  \
982      982         com.googlecode.jmockit:jmockit   
983      983         com.googlecode.jmockit:jmockit   
984      984         com.googlecode.jmockit:jmockit   
985      985         com.googlecode.jmockit:jmockit   
986      986         com.googlecode.jmockit:jmockit   
...      ...                                    ...   
11285  11285  org.powermock:powermock-module-junit4   
11286  11286  org.powermock:powermock-module-junit4   
11287  11287  org.powermock:powermock-module-junit4   
11288  11288  org.powermock:powermock-module-junit4   
11289  11289  org.powermock:powermock-module-junit4   

                                                   toLib  \
982                                  org.jmockit:jmockit   
983                                  org.jmockit:jmockit   
984                                  org.jmockit:jmockit   
985                                  org.jmockit:jmockit   
986                                  org.jmockit:jmockit   
...                                                  ...   
11285                           org.mockito:mockito-core   
11286                           org.mockito:mockito-core   
11287  org.springframework.boot:spring-boot-starter-test   
11288  org.springframework.boot:spring-boot-starter-test   
11289  org.springframework.boot:spring-boot-starter-test   

                                   repoName  \
982                    RUB-NDS_TLS-Attacker   
983                            apache_drill   
984                           DSpace_DSpace   
985                           DSpace_DSpace   
986                        javaee_glassfish   
...                                     ...   
11285              bonitasoft_bonita-engine   
11286  codecentric_chaos-monkey-spring-boot   
11287                         gazbert_bxbot   
11288                         gazbert_bxbot   
11289                         gazbert_bxbot   

                                                fileName  \
982                                              pom.xml   
983                                              pom.xml   
984                                              pom.xml   
985                              dspace-services/pom.xml   
986                                      nucleus/pom.xml   
...                                                  ...   
11285  bpm/bonita-core/bonita-parameter/bonita-parame...   
11286                   chaos-monkey-spring-boot/pom.xml   
11287                         bxbot-exchange-api/pom.xml   
11288                          bxbot-trading-api/pom.xml   
11289                       bxbot-domain-objects/pom.xml   

                                    startCommit  \
982    5c9d3ca9df2c61908d1f6bcce90df32bef3bf7a1   
983    8fe8f2cc25fa183d36fcb06ed533ffc9e501322c   
984    b810b99c89119f7338e168c0dffae50d822e06a7   
985    b810b99c89119f7338e168c0dffae50d822e06a7   
986    f010602d24f25aaa33fe1119d71ccfea2afecc2c   
...                                         ...   
11285  bb5bf02b270077645139dedcb064f27f92d6b740   
11286  faf715e32ba8d87da9e14ff4154af2940627689b   
11287  3c163531f279ffee3c66f2659ad20d6f464e9c86   
11288  3c163531f279ffee3c66f2659ad20d6f464e9c86   
11289  3c163531f279ffee3c66f2659ad20d6f464e9c86   

                                      endCommit  \
982    5c9d3ca9df2c61908d1f6bcce90df32bef3bf7a1   
983    8fe8f2cc25fa183d36fcb06ed533ffc9e501322c   
984    b810b99c89119f7338e168c0dffae50d822e06a7   
985    b810b99c89119f7338e168c0dffae50d822e06a7   
986    f010602d24f25aaa33fe1119d71ccfea2afecc2c   
...                                         ...   
11285  bb5bf02b270077645139dedcb064f27f92d6b740   
11286  faf715e32ba8d87da9e14ff4154af2940627689b   
11287  3c163531f279ffee3c66f2659ad20d6f464e9c86   
11288  3c163531f279ffee3c66f2659ad20d6f464e9c86   
11289  3c163531f279ffee3c66f2659ad20d6f464e9c86   

                                      startCommitChanges  \
982    +org.jmockit:jmockit\n-com.googlecode.jmockit:...

In [29]:
tolib_addition_test_df = original_df[
    original_df["toLib"].str.contains("rest-assured|robolectric|mock|assert")
]
tolib_addition_test_df

id                                fromLib  \
730      730             com.google.android:android   
732      732        com.google.android:android-test   
733      733        com.google.android:android-test   
981      981                 com.google.truth:truth   
982      982         com.googlecode.jmockit:jmockit   
...      ...                                    ...   
11282  11282  org.powermock:powermock-module-junit4   
11283  11283  org.powermock:powermock-module-junit4   
11284  11284  org.powermock:powermock-module-junit4   
11285  11285  org.powermock:powermock-module-junit4   
11286  11286  org.powermock:powermock-module-junit4   

                             toLib                              repoName  \
730    org.robolectric:android-all             serso_android-messengerpp   
732    org.robolectric:robolectric                     hoko_hoko-android   
733    org.robolectric:robolectric                          BUS-Team_bus   
981       org.assertj:assertj-core                      RyanSusana_elepy   
982            org.jmockit:jmockit                  RUB-NDS_TLS-Attacker   
...                            ...                                   ...   
11282     org.mockito:mockito-core              bonitasoft_bonita-engine   
11283     org.mockito:mockito-core              bonitasoft_bonita-engine   
11284     org.mockito:mockito-core              bonitasoft_bonita-engine   
11285     org.mockito:mockito-core              bonitasoft_bonita-engine   
11286     org.mockito:mockito-core  codecentric_chaos-monkey-spring-boot   

                                                fileName  \
730                                              pom.xml   
732                                              pom.xml   
733                                      android/pom.xml   
981                                        tests/pom.xml   
982                                              pom.xml   
...                                                  ...   
11282                   services/bonita-incident/pom.xml   
11283                                            pom.xml   
11284  services/bonita-scheduler/bonita-scheduler-imp...   
11285  bpm/bonita-core/bonita-parameter/bonita-parame...   
11286                   chaos-monkey-spring-boot/pom.xml   

                                    startCommit  \
730    d82e76b8ae0d90f59e0ce7a404855738fef6b3be   
732    c59afec9c1d13c21c118373a5e972805a2db5649   
733    cb50f8e47cbbf550e2421c9be86f14a96acfea2b   
981    6d5a498a1c92d79a1816ada5a8703ad41bf0544c   
982    5c9d3ca9df2c61908d1f6bcce90df32bef3bf7a1   
...                                         ...   
11282  bb5bf02b270077645139dedcb064f27f92d6b740   
11283  bb5bf02b270077645139dedcb064f27f92d6b740   
11284  bb5bf02b270077645139dedcb064f27f92d6b740   
11285  bb5bf02b270077645139dedcb064f27f92d6b740   
11286  faf715e32ba8d87da9e14ff4154af2940627689b   

                                      endCommit  \
730    d82e76b8ae0d90f59e0ce7a404855738fef6b3be   
732    c59afec9c1d13c21c118373a5e972805a2db5649   
733    cb50f8e47cbbf550e2421c9be86f14a96acfea2b   
981    6d5a498a1c92d79a1816ada5a8703ad41bf0544c   
982    5c9d3ca9df2c61908d1f6bcce90df32bef3bf7a1   
...                                         ...   
11282  bb5bf02b270077645139dedcb064f27f92d6b740   
11283  bb5bf02b270077645139dedcb064f27f92d6b740   
11284  bb5bf02b270077645139dedcb064f27f92d6b740   
11285  bb5bf02b270077645139dedcb064f27f92d6b740   
11286  faf715e32ba8d87da9e14ff4154af2940627689b   

                                      startCommitChanges  \
730    +org.robolectric:android-all\n+android:android...   
732    +org.easytesting:fest-assert-core\n+com.square...   
733    +org.roboguice:roboguice\n+org.robolectric:rob...   
981    +org.assertj:assertj-core\n-com.google.truth:t...   
982    +org.jmockit:jmockit\n-com.googlecode.jmockit:...   
...                                                  ...   
11282  +org.mockito:mockito-core\n-org.mockito:mockit...   
11283  +o

In [30]:
hard_remove_log_test_repos = (
    hard_remove_log_test_repos
    - set(fromlib_addition_test_df["repoName"].unique())
    - set(tolib_addition_test_df["repoName"].unique())
)
print(len(hard_remove_log_test_repos))

782


In [31]:
hard_remove_log_test_repos

{'1and1_cosmo',
 '42BV_jarb',
 '52North_SOS',
 '52North_series-rest-api',
 '55minutes_fiftyfive-wicket',
 'Accenture_mercury',
 'Alfresco_alfresco-sdk',
 'AmyAssist_Amy',
 'Appverse_appverse-web',
 'ArtificerRepo_artificer',
 'Asqatasun_Asqatasun',
 'Atmosphere_nettosphere',
 'AvanzaBank_astrix',
 'Azure_autorest-clientruntime-for-java',
 'Azure_azure-functions-java-worker',
 'Azure_azure-sdk-for-java',
 'BSC-RUS_AuTe-Framework',
 'Bammerbom_UltimateCore',
 'BatooOrg_BatooJPA',
 'Blazebit_blaze-persistence',
 'BorderTech_wcomponents',
 'BroadleafCommerce_BroadleafCommerce',
 'BrunoEberhard_minimal-j',
 'BrynCooke_cdi-unit',
 'BuycraftPlugin_BuycraftX',
 'CCI-MIT_XCoLab',
 'CUTR-at-USF_gtfs-realtime-validator',
 'CenterForOpenScience_cas-overlay',
 'ChargeTimeEU_Java-OCA-OCPP',
 'ClearTK_cleartk',
 'CloudSlang_score',
 'Cloudname_cloudname',
 'CloverETL_CloverETL-Engine',
 'CodeSolid_tutorials',
 'Cognifide_Slice',
 'Consonance_consonance',
 'CoreMedia_jangaroo-tools',
 'Dempsy_dempsy',

In [32]:
hard_remove_log_test_df = original_df[
    original_df["repoName"].isin(hard_remove_log_test_repos)
]
hard_remove_log_test_df

id          fromLib                             toLib  \
2          2          ant:ant                org.apache.ant:ant   
10        10          ant:ant                org.apache.ant:ant   
11        11          ant:ant                org.apache.ant:ant   
13        13          ant:ant                org.apache.ant:ant   
14        14          ant:ant                org.apache.ant:ant   
...      ...              ...                               ...   
14316  14316  xstream:xstream  com.thoughtworks.xstream:xstream   
14317  14317  xstream:xstream  com.thoughtworks.xstream:xstream   
14327  14327  xstream:xstream  com.thoughtworks.xstream:xstream   
14332  14332  xstream:xstream  com.thoughtworks.xstream:xstream   
14333  14333  xstream:xstream  com.thoughtworks.xstream:xstream   

                  repoName                                   fileName  \
2              teiid_teiid                                    pom.xml   
10           Top-Q_jsystem  jsystem-core-projects/jsystemCore/pom.xml   
11           Top-Q_jsystem                     jsystem-parent/pom.xml   
13     geoserver_geoserver                                src/pom.xml   
14     geoserver_geoserver                        src/release/pom.xml   
...                    ...                                        ...   
14316           apache_ace         ace-client-repository-impl/pom.xml   
14317           apache_ace    ace-client-repository-useradmin/pom.xml   
14327           bcdev_beam                                    pom.xml   
14332       apache_shindig   1.0.x-incubating/java/social-api/pom.xml   
14333       apache_shindig                   1.0.x-incubating/pom.xml   

                                    startCommit  \
2      204ea9614df7b99c46cc054877c5bea4d145b460   
10     69e5982631fb1a4932789e81ef19291780c8b05c   
11     69e5982631fb1a4932789e81ef19291780c8b05c   
13     8e51ce269eaab07cfa04b700e2dcb909133675ed   
14     8e51ce269eaab07cfa04b700e2dcb909133675ed   
...                                         ...   
14316  3df4226489929e6ab565f99481826acd9e7a0b4f   
14317  3df4226489929e6ab565f99481826acd9e7a0b4f   
14327  769b9c89c3173be4fd174e374f08c762c8005edd   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      endCommit  \
2      204ea9614df7b99c46cc054877c5bea4d145b460   
10     69e5982631fb1a4932789e81ef19291780c8b05c   
11     69e5982631fb1a4932789e81ef19291780c8b05c   
13     8e51ce269eaab07cfa04b700e2dcb909133675ed   
14     8e51ce269eaab07cfa04b700e2dcb909133675ed   
...                                         ...   
14316  3df4226489929e6ab565f99481826acd9e7a0b4f   
14317  3df4226489929e6ab565f99481826acd9e7a0b4f   
14327  769b9c89c3173be4fd174e374f08c762c8005edd   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      startCommitChanges  \
2                          +org.apache.ant:ant\n-ant:ant   
10                         +org.apache.ant:ant\n-ant:ant   
11                         +org.apache.ant:ant\n-ant:ant   
13                         +org.apache.ant:ant\n-ant:ant   
14                         +org.apache.ant:ant\n-ant:ant   
...                                                  ...   
14316  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14317  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14327  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14332  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14333  +com.thoughtworks.xstream:xstream\n-xstream:xs...   

                                        endCommitChanges  \
2                          +org.apache.ant:ant\n-ant:ant   
10                         +org.apache.ant:ant\n-ant:ant   
11                         +org.apache.ant:ant\n-ant:ant   
13                         +org.apache.ant:ant\n-ant:ant   
14                         +org.apache.ant:ant\n-ant:ant   
...                  

# Insight about number of pair fromLib-toLib per repo

In [34]:
migration_in_repo = dict(hard_remove_log_test_df["repoName"].value_counts())

In [35]:
len(list(filter(lambda x: migration_in_repo[x] > 1, migration_in_repo)))

414

In [36]:
repo_with_only_one_migration = list(
    filter(lambda x: migration_in_repo[x] == 1, migration_in_repo)
)
repo_with_only_one_migration

['gocd_go-plugins',
 'EmiteGWT_emite',
 'a466350665_smart',
 'thubbo_jmeter-plugins-for-apache-dubbo',
 'opennetworkinglab_onos-app-samples',
 'darkmi_rtspserver',
 'dianping_cat',
 'alibaba_Sentinel',
 'esl-client_esl-client',
 'mheath_adbcj',
 'menacher_java-game-server',
 'wfh45678_radar',
 'naver_ngrinder',
 'jboss-switchyard_components',
 'everit-org_json-schema',
 'wso2_wso2-synapse',
 'apifest_apifest',
 'uwolfer_gerrit-intellij-plugin',
 'ChargeTimeEU_Java-OCA-OCPP',
 'ibinti_bugvm',
 'mulesoft_mule-maven-plugin',
 'robovm_robovm',
 'gchq_Gaffer',
 'uwolfer_gerrit-rest-java-client',
 'apache_incubator-iotdb',
 'suninformation_ymateplatform',
 'SmartDataAnalytics_Sparqlify',
 'apache_karaf-cellar',
 'VueGWT_vue-gwt',
 'ata4_disunity',
 'arquillian_arquillian-core',
 'arquillian_arquillian-container-tomcat',
 'swri-robotics_bag-database',
 'zauberlabs_gnip4j',
 'dcm4che_dcm4chee-arc-light',
 'marytts_marytts',
 'apache_incubator-dlab',
 'joshsh_ripple',
 'kuali_kfs',
 'wooti_oned

In [37]:
len(repo_with_only_one_migration)

368

In [38]:
check = dict(original_df["repoName"].value_counts())
check

{'apache_camel': 1216,
 'keycloak_keycloak': 297,
 'Governance_rtgov': 226,
 'nuxeo_nuxeo': 194,
 'apache_usergrid': 191,
 'neo4j_neo4j': 183,
 'apache_knox': 180,
 'apache_hbase': 173,
 'apache_cxf': 166,
 'CamelCookbook_camel-cookbook-examples': 136,
 'Sage-Bionetworks_Synapse-Repository-Services': 130,
 'thorntail_thorntail': 121,
 'apache_james-project': 120,
 'sakaiproject_sakai': 111,
 'wicketstuff_core': 109,
 'Fabric3_fabric3-core': 107,
 'payara_Payara': 103,
 'opendaylight_controller': 93,
 'apache_hadoop': 89,
 'iluwatar_java-design-patterns': 87,
 'apache_flume': 85,
 'apache_syncope': 81,
 'Activiti_Activiti': 79,
 'Intel-bigdata_SSM': 77,
 'strongbox_strongbox': 72,
 'opencit_opencit': 69,
 'weld_core': 68,
 'kiegroup_kogito-runtimes': 66,
 'mulesoft_mule': 63,
 'infinispan_infinispan': 63,
 'MyCoRe-Org_mycore': 63,
 'apache_archiva': 62,
 'apiman_apiman': 61,
 'cucumber_cucumber-jvm': 59,
 'spring-projects_spring-boot': 58,
 'apache_storm': 58,
 'apache_servicecomb-java-

In [39]:
len(list(filter(lambda x: check[x] > 1, check)))

982

In [40]:
982 / 1480

0.6635135135135135

In [41]:
one_df = original_df[original_df["repoName"].isin(repo_with_only_one_migration)]
one_df

id                                         fromLib  \
84        84                                         asm:asm   
88        88                                         asm:asm   
89        89                                         asm:asm   
90        90                                         asm:asm   
93        93                                         asm:asm   
...      ...                                             ...   
14155  14155                                        rhino:js   
14160  14160                                        rhino:js   
14174  14174  ru.yandex.qatools.properties:properties-loader   
14270  14270                               velocity:velocity   
14288  14288                                     xalan:xalan   

                               toLib  \
84                   org.ow2.asm:asm   
88                   org.ow2.asm:asm   
89                   org.ow2.asm:asm   
90                   org.ow2.asm:asm   
93           org.ow2.asm:asm-commons   
...                              ...   
14155              org.mozilla:rhino   
14160              org.mozilla:rhino   
14174  ru.qatools.commons:properties   
14270   org.apache.velocity:velocity   
14288    net.sourceforge.saxon:saxon   

                                             repoName  \
84                                eclipse_sisu.inject   
88                                noctarius_Lightning   
89                                         oblac_jodd   
90                                    Cognifide_Slice   
93                                      hcoles_pitest   
...                                               ...   
14155                        happygiraffe_jslint4java   
14160                     picocontainer_picocontainer   
14174  yandex-qatools_hamcrest-pojo-matcher-generator   
14270                               mojohaus_webstart   
14288                                epimorphics_elda   

                                fileName  \
84       org.eclipse.sisu.inject/pom.xml   
88                               pom.xml   
89                               pom.xml   
90                    slice-core/pom.xml   
93                        pitest/pom.xml   
...                                  ...   
14155                            pom.xml   
14160        script/script-rhino/pom.xml   
14174  feature-matcher-generator/pom.xml   
14270      webstart-maven-plugin/pom.xml   
14288                elda-common/pom.xml   

                                    startCommit  \
84     01ca345463fdaee4d5d72293bb7679fb5bc815ec   
88     8ec806a0986990bd80e609b62ec055753fc97e6a   
89     c057522ed029f1b223207eb22ba54dbd80623bba   
90     c4cffec426469641b6652f9ffb1b5b34f7639329   
93     0d638d02c051094ce1e8532e83449735cf6ae2f4   
...                                         ...   
14155  86a643ff774b2462c0032b1feaeafb1dd09d6266   
14160  d1544d25ba23121ddf43073f62f7ca7330991550   
14174  cc7a474d3ea53164d28b4a1b23c25989982e7d03   
14270  e6d0944e489b7849c31957b2b41ca91044aeba65   
14288  e07982e05c1d9900aa91c87f28ec7bd0ba0a8f0f   

                                      endCommit  \
84     01ca345463fdaee4d5d72293bb7679fb5bc815ec   
88     8ec806a0986990bd80e609b62ec055753fc97e6a   
89     c057522ed029f1b223207eb22ba54dbd80623bba   
90     c4cffec426469641b6652f9ffb1b5b34f7639329   
93     0d638d02c051094ce1e8532e83449735cf6ae2f4   
...                                         ...   
14155  86a643ff774b2462c0032b1feaeafb1dd09d6266   
14160  d1544d25ba23121ddf43073f62f7ca7330991550   
14174  cc7a474d3ea53164d28b4a1b23c25989982e7d03   
14270  e6d0944e489b7849c31957b2b41ca91044aeba65   
14288  e07982e05c1d9900aa91c87f28ec7bd0ba0a8f0f   

                                      startCommitChanges  \
84                            +org.ow2.asm:asm\n-asm:asm   
88                            +org.ow2.asm:asm\n-asm:asm   
89                            +org.ow2.asm:asm\n-asm:asm   
90                            +org.ow2.asm:asm\n-asm:asm   
93     +org.ow2.asm:

In [42]:
one_df.to_csv("demo.csv", index=False)

In [43]:
sampled_df = one_df.sample(n=20, random_state=42)
sampled_df

id                                            fromLib  \
3878    3878                                 javax.servlet:jstl   
729      729                     com.github.sgroschupf:zkclient   
469      469                                  com.alibaba:dubbo   
11215  11215                          org.postgresql:postgresql   
1013    1013                                  com.h2database:h2   
6683    6683                             org.apache.derby:derby   
1412    1412                         com.sun.jersey:jersey-core   
2243    2243                          commons-lang:commons-lang   
3618    3618                          javax.mail:javax.mail-api   
2418    2418                          commons-lang:commons-lang   
7095    7095               org.apache.tomcat:tomcat-servlet-api   
884      884                          com.google.code.gson:gson   
3616    3616                          javax.json:javax.json-api   
1009    1009                                  com.h2database:h2   
3175    3175             io.dropwizard.metrics:metrics-graphite   
6562    6562                   org.apache.commons:commons-dbcp2   
84        84                                            asm:asm   
6782    6782                        org.apache.kafka:kafka_2.11   
13859  13859  org.springframework.security.oauth:spring-secu...   
1925    1925              commons-httpclient:commons-httpclient   

                                              toLib  \
3878                javax.servlet.jsp.jstl:jstl-api   
729                             com.101tec:zkclient   
469                          org.apache.dubbo:dubbo   
11215          org.mariadb.jdbc:mariadb-java-client   
1013                     mysql:mysql-connector-java   
6683                              com.h2database:h2   
1412        org.glassfish.jersey.core:jersey-client   
2243               org.apache.commons:commons-lang3   
3618                        com.sun.mail:javax.mail   
2418               org.apache.commons:commons-lang3   
7095                javax.servlet:javax.servlet-api   
884                            com.alibaba:fastjson   
3616                       org.glassfish:javax.json   
1009                     mysql:mysql-connector-java   
3175   io.micrometer:micrometer-registry-prometheus   
6562                            com.zaxxer:HikariCP   
84                                  org.ow2.asm:asm   
6782         org.springframework.kafka:spring-kafka   
13859                          io.jsonwebtoken:jjwt   
1925           org.apache.httpcomponents:httpclient   

                                       repoName  \
3878                      richfaces4_components   
729                    killme2008_Metamorphosis   
469                          apache_dubbo-admin   
11215                      FundRequest_platform   
1013                               JoeCao_qbike   
6683              wso2_ballerina-message-broker   
1412                  urbanairship_java-library   
2243                        JULIELab_jcore-base   
3618                        HubSpot_Singularity   
2418                            estatio_estatio   
7095                            cegeka_batchers   
884                            hippo-band_hippo   
3616                        stanfordnlp_CoreNLP   
1009                        nilzao_soapbox-race   
3175                  HotelsDotCom_data-highway   
6562                 plan-player-analytics_Plan   
84                          eclipse_sisu.inject   
6782                   daydreamdev_seconds-kill   
13859  gravitee-io_graviteeio-access-management   
1925                             EmiteGWT_emite   

                                                fileName  \
3878                                      parent/pom.xml   
729                                              pom.xml   
469                                              pom.xml   
11215                                       core/pom.xml   
1013                                          uc/pom.xml   
6683          

In [44]:
sampled_df.to_csv("sampled.csv", index=False)

In [45]:
(one_df["startCommit"] == one_df["endCommit"]).value_counts()

True     364
False      4
Name: count, dtype: int64

In [46]:
(sampled_df["startCommit"] == sampled_df["endCommit"]).value_counts()

True    20
Name: count, dtype: int64

In [56]:
original_df["num_migration"] = original_df.groupby("repoName")["repoName"].transform(
    "count"
)
original_df

id          fromLib                             toLib  \
0          0          ant:ant                org.apache.ant:ant   
1          1          ant:ant                org.apache.ant:ant   
2          2          ant:ant                org.apache.ant:ant   
7          7          ant:ant                org.apache.ant:ant   
8          8          ant:ant                org.apache.ant:ant   
...      ...              ...                               ...   
14329  14329  xstream:xstream  com.thoughtworks.xstream:xstream   
14330  14330  xstream:xstream  com.thoughtworks.xstream:xstream   
14331  14331  xstream:xstream  com.thoughtworks.xstream:xstream   
14332  14332  xstream:xstream  com.thoughtworks.xstream:xstream   
14333  14333  xstream:xstream  com.thoughtworks.xstream:xstream   

                       repoName                                  fileName  \
0           bobmcwhirter_drools                                   pom.xml   
1            wocommunity_wolips               woproject-ant-tasks/pom.xml   
2                   teiid_teiid                                   pom.xml   
7                 mulesoft_mule                        standalone/pom.xml   
8                 mulesoft_mule          distributions/standalone/pom.xml   
...                         ...                                       ...   
14329  kiegroup_kogito-runtimes                       drools-core/pom.xml   
14330        sakaiproject_sakai            gradebook/service/impl/pom.xml   
14331        sakaiproject_sakai      gradebook/service/sakai-pack/pom.xml   
14332            apache_shindig  1.0.x-incubating/java/social-api/pom.xml   
14333            apache_shindig                  1.0.x-incubating/pom.xml   

                                    startCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      endCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      startCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.apache.ant:ant\n-ant:ant   
...                                                  ...   
14329  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14330  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14331  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14332  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14333  +com.thoughtworks.xstream:xstream\n-xstream:xs...   

                                        endCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.a

In [57]:
original_df["num_migration"]

0         14
1          2
2          9
7         63
8         63
        ... 
14329     66
14330    111
14331    111
14332     19
14333     19
Name: num_migration, Length: 13382, dtype: int64

In [58]:
13382 / 1480

9.041891891891892

In [59]:
check = original_df.drop_duplicates(subset=["repoName"])
check.describe()

id     Category  num_migration
count   1480.000000  1480.000000    1480.000000
mean    4973.727703    29.189189       9.041892
std     4105.981888    26.246775      37.256701
min        0.000000     0.000000       1.000000
25%     1337.750000    11.000000       1.000000
50%     3975.500000    18.000000       2.000000
75%     7582.500000    39.000000       7.000000
max    14315.000000   115.000000    1216.000000

# Check file change

In [1]:
import pandas as pd

df = pd.read_csv("processed2.csv")
df.head()

id                         fromLib  \
0   3878              javax.servlet:jstl   
1    729  com.github.sgroschupf:zkclient   
2    469               com.alibaba:dubbo   
3  11215       org.postgresql:postgresql   
4   1013               com.h2database:h2   

                                  toLib                  repoName  \
0       javax.servlet.jsp.jstl:jstl-api     richfaces4_components   
1                   com.101tec:zkclient  killme2008_Metamorphosis   
2                org.apache.dubbo:dubbo        apache_dubbo-admin   
3  org.mariadb.jdbc:mariadb-java-client      FundRequest_platform   
4            mysql:mysql-connector-java              JoeCao_qbike   

         fileName                               startCommit  \
0  parent/pom.xml  82b5c6d243ac4b566ec9cfbf5cbe3a306481000d   
1         pom.xml  f6254da1916ec87ac12fd5da1469408e91aa320a   
2         pom.xml  995e06b547cb7554c99deb3129fca19b2c35aa8b   
3    core/pom.xml  101f86c7e7786dcfd0accdf4083f904224329fa5   
4      uc/pom.xml  ba1ba24af6b9a69ea6ed4d76ae61faf9191e4c0f   

                                  endCommit  \
0  82b5c6d243ac4b566ec9cfbf5cbe3a306481000d   
1  f6254da1916ec87ac12fd5da1469408e91aa320a   
2  995e06b547cb7554c99deb3129fca19b2c35aa8b   
3  101f86c7e7786dcfd0accdf4083f904224329fa5   
4  ba1ba24af6b9a69ea6ed4d76ae61faf9191e4c0f   

                                  startCommitChanges  \
0  +javax.servlet.jsp.jstl:jstl-api\n-javax.servl...   
1  +com.101tec:zkclient\n-com.github.sgroschupf:z...   
2  +org.apache.dubbo:dubbo\n+org.apache.dubbo:dub...   
3  +org.mariadb.jdbc:mariadb-java-client\n-org.po...   
4    +mysql:mysql-connector-java\n-com.h2database:h2   

                                    endCommitChanges  \
0  +javax.servlet.jsp.jstl:jstl-api\n-javax.servl...   
1  +com.101tec:zkclient\n-com.github.sgroschupf:z...   
2  +org.apache.dubbo:dubbo\n+org.apache.dubbo:dub...   
3  +org.mariadb.jdbc:mariadb-java-client\n-org.po...   
4    +mysql:mysql-connector-java\n-com.h2database:h2   

                          startCommitMessage  \
0  Updated jstl artifact in pom (RF-11077)\n   
1                  Upgrade zkclient to 0.3\n   
2  change dubbo dependency to 2.7-SNAPSHOT\n   
3                        switch to mariadb\n   
4       move to mysql and add trip service\n   

                            endCommitMessage startCommitTime endCommitTime  \
0  Updated jstl artifact in pom (RF-11077)\n      2011-06-19    2011-06-19   
1                  Upgrade zkclient to 0.3\n      2013-07-14    2013-07-14   
2  change dubbo dependency to 2.7-SNAPSHOT\n      2018-11-14    2018-11-14   
3                        switch to mariadb\n      2017-04-07    2017-04-07   
4       move to mysql and add trip service\n      2017-11-20    2017-11-20   

   Category    repoOwner  repoSplitName  \
0        66   richfaces4     components   
1        15   killme2008  Metamorphosis   
2        16       apache    dubbo-admin   
3        31  FundRequest       platform   
4        31       JoeCao          qbike   

                                prev_commit  \
0  0a002564ee00679be07b70f4902e82e1fa8bb45a   
1  efb075b1bed3368b8ff961656c2d9dc691c8a6e4   
2  80013ffc38b3d737bd891a2b574ef75a93450bd9   
3                                       NaN   
4  3c876617d0e3506205fd9d02dbf050a7565f796c   

                                          diff_files  
0  {'Added': [], 'Modified': [], 'Deleted': [], '...  
1  {'Added': [], 'Modified': [], 'Deleted': [], '...  
2  {'Added': [], 'Modified': ['dubbo-admin-backen...  
3                                                NaN  
4  {'Added': ['intention/src/main/java/club/newte...

In [ ]:
import subprocess
def read_file_in_commit(repo_dir, rev_path, commit_hash):
    

In [ ]:
for i in range(len(df)):
    diff_files = df["diff_files"].iloc[i]    
    if pd.isna(diff_files):
        continue
    diff_files = eval(diff_files)
    if diff_files["Renamed"]:
        

[('intention/src/main/java/club/newtech/qbike/domain/core/vo/User.java', 'trip/src/main/java/club/newtech/qbike/trip/domain/core/vo/Driver.java'), ('intention/src/main/java/club/newtech/qbike/domain/core/service/UserRibbonHystrixService.java', 'trip/src/main/java/club/newtech/qbike/trip/infrastructure/UserRibbonHystrixApi.java')]
[('gravitee-am-management-api/gravitee-am-management-api-handlers/gravitee-am-management-api-handlers-oauth2/src/main/java/io/gravitee/am/management/handlers/oauth2/WebMvcConfiguration.java', 'gravitee-am-management-api/gravitee-am-management-api-handlers/gravitee-am-management-api-handlers-admin/src/main/java/io/gravitee/am/management/handlers/admin/WebMvcConfiguration.java'), ('gravitee-am-management-api/gravitee-am-management-api-handlers/gravitee-am-management-api-handlers-oauth2/src/main/java/io/gravitee/am/management/handlers/oauth2/authentication/CustomSavedRequestAwareAuthenticationSuccessHandler.java', 'gravitee-am-management-api/gravitee-am-managemen

In [1]:
import pandas as pd

df = pd.read_csv("methods_20.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80 entries, 0 to 79
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   id                  80 non-null     int64 
 1   fromLib             80 non-null     object
 2   toLib               80 non-null     object
 3   repoName            80 non-null     object
 4   fileName            80 non-null     object
 5   startCommit         80 non-null     object
 6   endCommit           80 non-null     object
 7   startCommitChanges  80 non-null     object
 8   endCommitChanges    75 non-null     object
 9   startCommitMessage  80 non-null     object
 10  endCommitMessage    80 non-null     object
 11  startCommitTime     80 non-null     object
 12  endCommitTime       80 non-null     object
 13  Category            80 non-null     int64 
 14  repoOwner           80 non-null     object
 15  repoSplitName       80 non-null     object
 16  prev_commit         80 non-n

In [2]:
df_simplified = df.drop(columns=["diff_files"])
df_simplified.to_csv("methods_20_simplified.csv", index=False)

# New check

In [34]:
import pandas as pd

df = pd.read_parquet("official_original.parquet")
df

id          fromLib                             toLib  \
0          0          ant:ant                org.apache.ant:ant   
1          1          ant:ant                org.apache.ant:ant   
2          2          ant:ant                org.apache.ant:ant   
7          7          ant:ant                org.apache.ant:ant   
8          8          ant:ant                org.apache.ant:ant   
...      ...              ...                               ...   
14329  14329  xstream:xstream  com.thoughtworks.xstream:xstream   
14330  14330  xstream:xstream  com.thoughtworks.xstream:xstream   
14331  14331  xstream:xstream  com.thoughtworks.xstream:xstream   
14332  14332  xstream:xstream  com.thoughtworks.xstream:xstream   
14333  14333  xstream:xstream  com.thoughtworks.xstream:xstream   

                       repoName                                  fileName  \
0           bobmcwhirter_drools                                   pom.xml   
1            wocommunity_wolips               woproject-ant-tasks/pom.xml   
2                   teiid_teiid                                   pom.xml   
7                 mulesoft_mule                        standalone/pom.xml   
8                 mulesoft_mule          distributions/standalone/pom.xml   
...                         ...                                       ...   
14329  kiegroup_kogito-runtimes                       drools-core/pom.xml   
14330        sakaiproject_sakai            gradebook/service/impl/pom.xml   
14331        sakaiproject_sakai      gradebook/service/sakai-pack/pom.xml   
14332            apache_shindig  1.0.x-incubating/java/social-api/pom.xml   
14333            apache_shindig                  1.0.x-incubating/pom.xml   

                                    startCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      endCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      startCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.apache.ant:ant\n-ant:ant   
...                                                  ...   
14329  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14330  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14331  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14332  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14333  +com.thoughtworks.xstream:xstream\n-xstream:xs...   

                                        endCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.a

In [35]:
df["repoName"].nunique()

1480

In [36]:
df["startCommit"].nunique()

3038

In [37]:
df["endCommit"].nunique()

3029

In [42]:
df["repo_start_end_commit"] = df.apply(lambda row: row["repoName"] + "__" + row["startCommit"] + "__" + row["endCommit"], axis=1)
df

id          fromLib                             toLib  \
0          0          ant:ant                org.apache.ant:ant   
1          1          ant:ant                org.apache.ant:ant   
2          2          ant:ant                org.apache.ant:ant   
7          7          ant:ant                org.apache.ant:ant   
8          8          ant:ant                org.apache.ant:ant   
...      ...              ...                               ...   
14329  14329  xstream:xstream  com.thoughtworks.xstream:xstream   
14330  14330  xstream:xstream  com.thoughtworks.xstream:xstream   
14331  14331  xstream:xstream  com.thoughtworks.xstream:xstream   
14332  14332  xstream:xstream  com.thoughtworks.xstream:xstream   
14333  14333  xstream:xstream  com.thoughtworks.xstream:xstream   

                       repoName                                  fileName  \
0           bobmcwhirter_drools                                   pom.xml   
1            wocommunity_wolips               woproject-ant-tasks/pom.xml   
2                   teiid_teiid                                   pom.xml   
7                 mulesoft_mule                        standalone/pom.xml   
8                 mulesoft_mule          distributions/standalone/pom.xml   
...                         ...                                       ...   
14329  kiegroup_kogito-runtimes                       drools-core/pom.xml   
14330        sakaiproject_sakai            gradebook/service/impl/pom.xml   
14331        sakaiproject_sakai      gradebook/service/sakai-pack/pom.xml   
14332            apache_shindig  1.0.x-incubating/java/social-api/pom.xml   
14333            apache_shindig                  1.0.x-incubating/pom.xml   

                                    startCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      endCommit  \
0      0504ef4b2349dd7edc4d4991d29f688dc15da939   
1      1d02c267e787b0c97ef9a3fb3b63da33e8f3c6d9   
2      204ea9614df7b99c46cc054877c5bea4d145b460   
7      4a47dcdc23f8800049e1aabd29ea5d618b0ac7dd   
8      55f25f71b131ccac1c7e7c2cba99b330759ad574   
...                                         ...   
14329  ae5bf7e576bce9bc7c1871b323dc81bbded46b55   
14330  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14331  d32f1ecd4a80a1f4b1ccda1763c6d7f0f5462584   
14332  f6cabfb61fbbc6a6afbda13b39189c005c681967   
14333  f6cabfb61fbbc6a6afbda13b39189c005c681967   

                                      startCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.apache.ant:ant\n-ant:ant   
...                                                  ...   
14329  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14330  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14331  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14332  +com.thoughtworks.xstream:xstream\n-xstream:xs...   
14333  +com.thoughtworks.xstream:xstream\n-xstream:xs...   

                                        endCommitChanges  \
0      +org.apache.ant:ant\n+org.apache.ant:ant-apach...   
1      +org.apache.ant:ant\n+org.apache.ant:ant-junit...   
2                          +org.apache.ant:ant\n-ant:ant   
7                          +org.apache.ant:ant\n-ant:ant   
8                          +org.a

In [43]:
df["repo_start_end_commit"].nunique()

3189

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13382 entries, 0 to 14333
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   id                     13382 non-null  int64 
 1   fromLib                13382 non-null  object
 2   toLib                  13382 non-null  object
 3   repoName               13382 non-null  object
 4   fileName               13382 non-null  object
 5   startCommit            13382 non-null  object
 6   endCommit              13382 non-null  object
 7   startCommitChanges     13382 non-null  object
 8   endCommitChanges       13380 non-null  object
 9   startCommitMessage     13380 non-null  object
 10  endCommitMessage       13380 non-null  object
 11  startCommitTime        13382 non-null  object
 12  endCommitTime          13382 non-null  object
 13  Category               13382 non-null  int64 
 14  repoOwner              13382 non-null  object
 15  repoSplitName          1

In [47]:
for i in range(len(df) - 1):
    for j in range(i + 1, len(df)):
        if df.iloc[i, 16] == df.iloc[j, 16] and df.iloc[i, 17] != df.iloc[j, 17]:
            print(i, j)

90 92
90 125
90 126
91 92
91 125
91 126
92 122
92 123
92 124
122 125
122 126
123 125
123 126
124 125
124 126
183 184
183 324
183 10634
184 323
184 10633
189 190
189 10638
190 10637
200 202
200 203
200 221
200 222
200 257
200 258
200 322
201 202
201 203
201 221
201 222
201 257
201 258
201 322
202 219
202 220
202 255
202 256
202 321
203 219
203 220
203 255
203 256
203 321
204 205
204 225
204 271
204 330
205 224
205 270
205 329
219 221
219 222
219 257
219 258
219 322
220 221
220 222
220 257
220 258
220 322
221 255
221 256
221 321
222 255
222 256
222 321
224 225
224 271
224 330
225 270
225 329
255 257
255 258
255 322
256 257
256 258
256 322
257 321
258 321
270 271
270 330
271 329
321 322
323 324
323 10634
324 10633
329 330
341 342
341 343
342 343


KeyboardInterrupt: 

In [49]:
df.to_csv("official_added_id.csv", index=False)

In [50]:
unique = df.groupby("repo_start_end_commit").apply(lambda x: x.sample(1, random_state=42))
unique

/tmp/ipykernel_349975/1081030082.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  unique = df.groupby("repo_start_end_commit").apply(lambda x: x.sample(1, random_state=42))


id  \
repo_start_end_commit                                             
1and1_cosmo__f8207e3129d721db5026cc9ea2a8ce58db... 10470  10470   
42BV_jarb__b193a20c23a862be0ec7abccf74b83f25979... 2316    2316   
52North_SOS__1dcf0c8a359f6673272df58d5419af3297... 1667    1667   
52North_SOS__520cc60fd93dbfb6e48d75f23765d72556... 14029  14029   
52North_SOS__a46610ab2c02a34e54e5a9985b96229655... 14122  14122   
...                                                         ...   
zkoss_zk__21a1ed83f0ec879f6ef973c82d51e147be208... 14152  14152   
zkoss_zk__555a531592fbb12e2621577f997545c076987... 7127    7127   
zkoss_zk__d283f0945a035fb61281c31ce3b13690aae34... 7117    7117   
zolyfarkas_spf4j__d321f19c34d5ee9fc91114d012553... 13786  13786   
zolyfarkas_spf4j__f5afbdf7dd26acf089e44f30c24e9... 13789  13789   

                                                                                       fromLib  \
repo_start_end_commit                                                                            
1and1_cosmo__f8207e3129d721db5026cc9ea2a8ce58db... 10470  org.mariadb.jdbc:mariadb-java-client   
42BV_jarb__b193a20c23a862be0ec7abccf74b83f25979... 2316              commons-lang:commons-lang   
52North_SOS__1dcf0c8a359f6673272df58d5419af3297... 1667                 com.vividsolutions:jts   
52North_SOS__520cc60fd93dbfb6e48d75f23765d72556... 14029                org.xerial:sqlite-jdbc   
52North_SOS__a46610ab2c02a34e54e5a9985b96229655... 14122                 postgresql:postgresql   
...                                                                                        ...   
zkoss_zk__21a1ed83f0ec879f6ef973c82d51e147be208... 14152                              rhino:js   
zkoss_zk__555a531592fbb12e2621577f997545c076987... 7127         org.codehaus.groovy:groovy-all   
zkoss_zk__d283f0945a035fb61281c31ce3b13690aae34... 7117                      org.beanshell:bsh   
zolyfarkas_spf4j__d321f19c34d5ee9fc91114d012553... 13786                org.slf4j:slf4j-simple   
zolyfarkas_spf4j__f5afbdf7dd26acf089e44f30c24e9... 13789                org.slf4j:slf4j-simple   

                                                                                     toLib  \
repo_start_end_commit                                                                        
1and1_cosmo__f8207e3129d721db5026cc9ea2a8ce58db... 10470         org.postgresql:postgresql   
42BV_jarb__b193a20c23a862be0ec7abccf74b83f25979... 2316   org.apache.commons:commons-lang3   
52North_SOS__1dcf0c8a359f6673272df58d5419af3297... 1667      org.locationtech.jts:jts-core   
52North_SOS__520cc60fd93dbfb6e48d75f23765d72556... 14029   javax.servlet:javax.servlet-api   
52North_SOS__a46610ab2c02a34e54e5a9985b96229655... 14122         org.postgresql:postgresql   
...                                                                                    ...   
zkoss_zk__21a1ed83f0ec879f6ef973c82d51e147be208... 14152                 org.mozilla:rhino   
zkoss_zk__555a531592fbb12e2621577f997545c076987... 7127                org.zkoss.zk:zuljsp   
zkoss_zk__d283f0945a035fb61281c31ce3b13690aae34... 7117    org.apache-extras.beanshell:bsh   
zolyfarkas_spf4j__d321f19c34d5ee9fc91114d012553... 13786        org.spf4j:spf4j-slf4j-test   
zolyfarkas_spf4j__f5afbdf7dd26acf089e44f30c24e9... 13789        org.spf4j:spf4j-slf4j-test   

                                                                  repoName  \
repo_start_end_commit                                                        
1and1_cosmo__f8207e3129d721db5026cc9ea2a8ce58db... 10470       1and1_cosmo   
42BV_jarb__b193a20c23a862be0ec7abccf74b83f25979... 2316          42BV_jarb   
52North_SOS__1dcf0c8a359f6673272df58d5419af3297... 1667        52North_SOS   
52North_SOS__520cc60fd93dbfb6e48d75f23765d72556... 14029       52North_SOS   
52North_SOS__a46610ab2c02a34e54e5a9985b96229655... 14122       52North_SOS   
...                                                                    ...   
zkoss_zk__21a1ed83f0ec879f6ef973c82d51e147be208... 14152          z

In [51]:
unique.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 3189 entries, ('1and1_cosmo__f8207e3129d721db5026cc9ea2a8ce58db34fea5__f8207e3129d721db5026cc9ea2a8ce58db34fea5', 10470) to ('zolyfarkas_spf4j__f5afbdf7dd26acf089e44f30c24e9122465f9d9b__f5afbdf7dd26acf089e44f30c24e9122465f9d9b', 13789)
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   id                     3189 non-null   int64 
 1   fromLib                3189 non-null   object
 2   toLib                  3189 non-null   object
 3   repoName               3189 non-null   object
 4   fileName               3189 non-null   object
 5   startCommit            3189 non-null   object
 6   endCommit              3189 non-null   object
 7   startCommitChanges     3189 non-null   object
 8   endCommitChanges       3187 non-null   object
 9   startCommitMessage     3188 non-null   object
 10  endCommitMessage       3188 non-null   object
 11  startCommitTi

In [52]:
sampled = unique.sample(n=50, random_state=42)
sampled

id  \
repo_start_end_commit                                             
apache_syncope__6cd08210d786af216a216e88e2f8dfa... 7989    7989   
apache_storm__3503dcea62c9bb9d004388773705ad362... 9649    9649   
apache_incubator-iotdb__f46aeb17e5b8734eb4887f7... 10237  10237   
SeqWare_seqware__289947f7719911a70d398c3df5a079... 7828    7828   
appfuse_appfuse__e6157cf6dcdc46521b8b1d640678a2... 6660    6660   
apache_camel__1e9f3ed611a758ffd49be254f7bf38d15... 11270  11270   
apache_camel__05c872497215f7f1b87d75f7087a703f8... 5600    5600   
emmanuelbernard_hibernate-core-ogm__f0755725c91... 2840    2840   
Flipkart_phantom__1fdac2a76bbd0cd5757adcbdccb9b... 12987  12987   
jbehave_jbehave-core__674ad3cdb9b7f40f95d855b0f... 2215    2215   
zalando_problem__1b987b88ecb5cc2c8df58ac8eda188... 4561    4561   
codenvy_codenvy__8a0a2026919db64ed2a49c73f7950a... 5359    5359   
RBGKew_Reconciliation-and-Matching-Framework__b... 8708    8708   
hibernate_hibernate-search__1c06d96a31af8cf8304... 11468  11468   
eclipse_microprofile-jwt-auth__d645c72925e3fe22... 3435    3435   
bmwcarit_joynr__787788996d6bd0391f1fc7b937460ca... 3268    3268   
jadler-mocking_jadler__2d6d47d2f90174e874511dfc... 1872    1872   
wooti_onedrive-java-client__0b7ee1a476e4a315bfb... 9446    9446   
TremoloSecurity_OpenUnison__0ae5c4d88bf6f65066e... 5691    5691   
apache_helix__308c2cc6f45d9c942d9b4a8f52e4cf534... 713      713   
richfaces4_dev-examples__9d194036e274f58b8fe5d1... 3891    3891   
sakaiproject_sakai__e26cef54bb0714ad5a38790a1a4... 2419    2419   
Wimmics_corese__936a23e22cf1abaff09a49026927697... 5729    5729   
RoboZonky_robozonky__6eebc8c0e3a328dcb94c99227d... 602      602   
ikasanEIP_ikasan__3a0116562e1b09d7bd803c8b6e0e7... 6153    6153   
ForgeRock_openidm-community-edition__43689602ee... 3249    3249   
apache_incubator-pinot__58791127a539c0b8db19c87... 10308  10308   
teiid_teiid__b710143466a04535075cd4c9fb56894f53... 569      569   
codenjoyme_codenjoy__9ca86c7a406eddfc69bb688fd8... 13924  13924   
structr_structr__569c38864cc772fddee47ff538c883... 11426  11426   
apache_isis__9cb3fc15b62b5ee44a02101a0184f77119... 9889    9889   
frenchc_jetm__2770a0899dff042abec0d9afe8141a1b5... 60        60   
afimb_chouette__45aaa983c274e2b6ab2f54af790109d... 6608    6608   
OpenNMS_opennms__c1d8968c36c055ff523726375f912f... 9786    9786   
deegree_deegree3__39c5ef900298dd2fe942fd60c3293... 14107  14107   
apache_struts__e990fbb751af3ed83123ca71bd90b76e... 2508    2508   
JakduK_jakduk-api__e586faf9633439ea6bd460066125... 4294    4294   
Intel-bigdata_SSM__049e7d27bea13d4254baccf49401... 8422    8422   
dana-i2cat_opennaas__50f7e0e5f6656020c0b0ec110a... 932      932   
ikasanEIP_ikasan__710d5c05b36b5fc464af42c417331... 2242    2242   
bcdev_beam__f06a705633f2f4d4c41c63237d76d7550bb... 4193    4193   
ikasanEIP_ikasan__54f8754cb7283742ec844e63cc045... 6074    6074   
vmware_hillview__5d417ba911779677ba3b11f77940a2... 3204    3204   
littlefishercoder_littlefisher-system__05551a5c... 11145  11145   
NukkitX_Nukkit__542354be77c337f5749a37928177303... 4234    4234   
iyzico_iyzipay-java__4cb3ff7df2a0e21bfbaf942bd9... 661      661   
nuxeo_nuxeo__bc0afb6481e5328e317a92f6571bb54698... 7683    7683   
krasserm_ipf__1a49df279f08860e6e512b2393ba7e23a... 3251    3251   
apache_directory-server__a17bab9b77162a4d44b7d7... 6636    6636   
ff4j_ff4j__f12d3871afd0645aeb354e5a91f1b8047687... 8404    8404   

                                                                                                 fromLib  \
repo_start_end_commit                                                                                      
apache_syncope__6cd08210d786af216a216e88e2f8dfa... 7989               org.codehaus.jackson:jackson-jaxrs   
apache_storm__3503dcea62c9bb9d004388773705ad362... 9649                        org.hamcrest:hamcrest-all   
apache_incubator-iotdb__f46aeb17e5b8734eb4887f7... 10237                                   org.json:json   
SeqWare_seqware__289947f7719911a70d3

In [54]:
sampled.to_csv("sampled_50.csv", index=False)